### PR TITLE
Introduce new overloads for ClientYamlTestExecutionContext and deprecate old ones

### PR DESCRIPTION
**Pull request by ldematte**
Smaller preliminary step extracted from https://github.com/elastic/elasticsearch/pull/103311/ to introduce a signature change without breaking submodule update in serverless.

The signature change is a bit larger - not only `Version` -> `Set<String>` but also 2 other changes that we now will be needed. Doing them in one go to minimize churn.

### CODE REVIEW

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/ESClientYamlSuiteTestCase.java [hunk 1]
```diff
@@ -190,6 +194,22 @@ public void initAndResetContext() throws Exception {
...
      * Create the test execution context. Can be overwritten in sub-implementations of the test if the context needs to be modified.
      */
+    protected ClientYamlTestExecutionContext createRestTestExecutionContext(
+        ClientYamlTestCandidate clientYamlTestCandidate,
+        ClientYamlTestClient clientYamlTestClient,
+        final Set<String> nodesVersions,
+        final TestFeatureService testFeatureService,
+        final Set<String> osList
+    ) {
+        return createRestTestExecutionContext(
```
[the remark this answers is not in the document]
Comment (ldematte) [reply test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/ESClientYamlSuiteTestCase.java:204]: Yes it is. In the serverless codebase, which knows only about the older overload, we overload the older/deprecated one.
So in order to keep working as before in the (short) transition, we have `ESClientYamlSuiteTestCase` -> (new) `createRestTestExecutionContext` -> (deprecated - overloaded) `createRestTestExecutionContext`

**Comment by ldematte** [comment]
@elasticmachine update branch

**Comment by ldematte** [comment]
@elasticmachine update branch

**Review by mosche — APPROVED** [review]
LGTM 👍

**Pull request merged**